### PR TITLE
Move frame_strings from Copter to Motors

### PR DESCRIPTION
**Pull request by kd0aij**
also adds frame_type strings

moves the get_frame_string method into the AP_Motors subclasses along with _last_frame_class and type, renaming them to _active_frame_class/type
the get_frame_string method is pure virtual and overridden by every subclass

When testing this, I noticed that copter tailsitter setup in quadplane was susceptible to configuration errors involving frame classes not supported by AP_MotorsMatrix;  these now result in a call to config_error.

Also, the gcs().send_text calls in QuadPlane::setup occur before the GCS is connected; changed these to ::printf so they always show up on the debug console.

Also added a quadplane frame class/type message to the log similar to Copter's.

### CODE REVIEW

File: ArduCopter/GCS_Copter.cpp [hunk 1]
```diff
@@ -9,7 +9,7 @@ uint8_t GCS_Copter::sysid_this_mav() const
 
 const char* GCS_Copter::frame_string() const
 {
-    return copter.get_frame_string();
+    return copter.motors->get_frame_string();
```
Comment (peterbarker) [inline ArduCopter/GCS_Copter.cpp:12]: We could use AP::motors() here.  I'm not sure its worth it, but I wonder if it might allow us to move code up at some stage.

Comment (peterbarker) [reply ArduCopter/GCS_Copter.cpp:12]: Also, this looks like a pointless method - could just fold the call into the user.

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -730,10 +738,10 @@ bool QuadPlane::setup(void)
     }
     AP_Param::load_object_from_eeprom(loiter_nav, loiter_nav->var_info);
 
-    motors->init((AP_Motors::motor_frame_class)frame_class.get(), (AP_Motors::motor_frame_type)frame_type.get());
+    motors->initialise(motor_class, motor_type);
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:741]: Why the name change here?  `init(...)` is pretty traditional.

File: ArduCopter/Log.cpp [hunk 3]
```diff
@@ -617,7 +617,7 @@ const struct LogStructure Copter::log_structure[] = {
 void Copter::Log_Write_Vehicle_Startup_Messages()
 {
     // only 200(?) bytes are guaranteed by AP_Logger
-    logger.Write_MessageF("Frame: %s", get_frame_string());
+    logger.Write_MessageF("Frame: %s, Type: %s", motors->get_frame_string(), motors->get_type_string());
```
Comment (peterbarker) [inline ArduCopter/Log.cpp:620]: Do you think "Layout" might be better than "type"?

File: ArduPlane/quadplane.cpp [hunk 4]
```diff
@@ -637,7 +636,9 @@ bool QuadPlane::setup(void)
       that the objects don't affect the vehicle unless enabled and
       also saves memory when not in use
      */
-    motor_class = (enum AP_Motors::motor_frame_class)frame_class.get();
+    motor_class = (AP_Motors::motor_frame_class)frame_class.get();
+    motor_type  = (AP_Motors::motor_frame_type)frame_type.get();
+    bool coptar_tailsitter_supported = true;
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:641]:
```suggestion
    bool copter_tailsitter_supported = true;
```

File: ArduPlane/quadplane.cpp [hunk 5]
```diff
@@ -688,8 +692,12 @@ bool QuadPlane::setup(void)
     } else {
         // this is a copter tailsitter with motor layout specified by frame_class and frame_type
         // tilting motors are not supported (tiltrotor control variables are ignored)
+        if (!coptar_tailsitter_supported) {
+            AP_BoardConfig::config_error("unsupported copterTS class: %u", motor_class);
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:696]:
```suggestion
            AP_BoardConfig::config_error("Unsupported copterTS class: %u", motor_class);
```

File: ArduPlane/quadplane.cpp [hunk 6]
```diff
@@ -730,10 +738,10 @@ bool QuadPlane::setup(void)
     }
     AP_Param::load_object_from_eeprom(loiter_nav, loiter_nav->var_info);
 
-    motors->init((AP_Motors::motor_frame_class)frame_class.get(), (AP_Motors::motor_frame_type)frame_type.get());
+    motors->initialise(motor_class, motor_type);
 
     if (!motors->initialised_ok()) {
-        AP_BoardConfig::config_error("unknown Q_FRAME_TYPE %u", (unsigned)frame_type.get());
+        AP_BoardConfig::config_error("init failed: class: %u type: %u", motor_class, motor_type);
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:744]: We do appear to have a valid motors object here, so shouldn't at least one of the string-getting methods work here?!

File: libraries/AP_Motors/AP_Motors_Class.h [hunk 7]
```diff
@@ -45,6 +45,43 @@ class AP_Motors {
         MOTOR_FRAME_DODECAHEXA = 12,
         MOTOR_FRAME_HELI_QUAD = 13,
     };
+
+    // return string corresponding to frame_class
+    virtual const char* get_frame_string()
+    {
+        switch (_active_frame_class) {
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_Motors_Class.h:52]: This is that if _active_frame_class corresponds to the allocated class then we *don't need* this switch statement.  There's just a method which returns the string.

This has the advantage of keeping things more self-contained so if you add a frame everything's in the one place.

File: libraries/AP_Motors/AP_Motors_Class.h [hunk 8]
```diff
@@ -156,7 +237,12 @@ class AP_Motors {
     virtual void        set_update_rate( uint16_t speed_hz ) { _speed_hz = speed_hz; }
 
     // init
-    virtual void        init(motor_frame_class frame_class, motor_frame_type frame_type) = 0;
+    void                initialise(motor_frame_class frame_class, motor_frame_type frame_type)
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_Motors_Class.h:240]: Ah.  We should probably change the name in this class rather than the externally visibly one.  "_init" is sometimes used :-)  Or "init_subclass"

**Review by peterbarker — COMMENTED** [review]
We could check with @rmackay9 to see if he's happy with the concept of doing away with the switch statements and having just the simple method call instead.

File: libraries/AP_Motors/AP_Motors_Class.h (hunk 8 above)
Comment (kd0aij) [reply libraries/AP_Motors/AP_Motors_Class.h:240]: I did it this way to avoid having to change every subclass. There are only 2 or 3 external usages of the old init() vs. 6 subclasses and 12 source files involved in overriding it.
But I see now that the init method is public in all the subclassses, so making it private would mean touching all of those files anyway. Guess I'll go back to renaming init to a private _init and edit all of the subclasses :(

File: libraries/AP_Motors/AP_Motors_Class.h (hunk 7 above)
Comment (kd0aij) [reply libraries/AP_Motors/AP_Motors_Class.h:52]: I don't see how to avoid a switch statement in MotorsMatrix, since it supports multiple frame classes and types

File: ArduPlane/quadplane.cpp (hunk 6 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:744]: I was thinking the class, type or both might be out of range of the respective enum, and in that case the integer param value would be more appropriate to report.

File: ArduPlane/quadplane.cpp (hunk 4 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:641]: wow, my eyes are getting worse :(
took forever to find the difference between those 2 strings

**Comment by peterbarker** [comment]
On Sun, 8 Nov 2020, Mark Whitehorn wrote:

> +    // return string corresponding to frame_class
> +    virtual const char* get_frame_string()
> +    {
> +        switch (_active_frame_class) {
> 
> I don't see how to avoid a switch statement in MotorsMatrix, since it
> supports multiple frame classes and types

Yes, it will need a switch.   But only on the types/classes that it 
supports.

**Comment by peterbarker** [comment]
On Sun, 8 Nov 2020, Mark Whitehorn wrote:

> 
> >
>      if (!motors->initialised_ok()) {
> -        AP_BoardConfig::config_error("unknown Q_FRAME_TYPE %u", (unsigned)f
> rame_type.get());
> +        AP_BoardConfig::config_error("init failed: class: %u type: %u", mot
> or_class, motor_type);
> 
> I was thinking the class, type or both might be out of range of the
> respective enum, and in that case the integer param value would be more
> appropriate to report.

You're in a block gated b the result code of a method call on an allocated 
object.  You can ask it what went wrong.

**Comment by kd0aij** [comment]
@rmackay9 Do you think that rearranging the get_frame_string approach in this way is worth doing?
It has the potential advantage of localizing the string values to the relevant motor class, reducing maintenance concerns and making the strings available to all vehicles.

File: libraries/AP_Motors/AP_MotorsSingle.cpp [hunk 9]
```diff
@@ -23,22 +23,27 @@ extern const AP_HAL::HAL& hal;
 // init
 void AP_MotorsSingle::init(motor_frame_class frame_class, motor_frame_type frame_type)
 {
-    // make sure 6 output channels are mapped
-    for (uint8_t i = 0; i < 6; i++) {
-        add_motor_num(CH_1 + i);
-    }
+    set_initialised_ok(false);
+    if (frame_class == MOTOR_FRAME_SINGLE) {
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_MotorsSingle.cpp:27]: If this isn't the case, why were we called?

I was expecting a simple `const char * get_frame_class() { return "Single"; }` in the header...

Comment (kd0aij) [reply libraries/AP_Motors/AP_MotorsSingle.cpp]: I didn't think to question why it was written this way, but I agree that for this subclass, it's better to assign the string in the header, and revert the cpp to its original state.

**Comment by peterbarker** [comment]
Yep, that's more like what I had in mind :-)

We're now calling methods on the motors object - we'd best see what happens if you don't have them defined properly - is it possible to get a method-on-undef problem?

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: gone

File: ArduCopter/Log.cpp (hunk 3 above)
Comment (kd0aij) [reply ArduCopter/Log.cpp:620]: Wouldn't that conflict with a lot of user-facing stuff: https://ardupilot.org/copter/docs/frame-type-configuration.html?highlight=frame%20type

File: ArduPlane/quadplane.cpp (hunk 5 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:696]: fixed

File: ArduPlane/quadplane.cpp (hunk 6 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:744]: initialized_ok is a boolean, it doesn't know why the failure occurred

**Comment by peterbarker** [comment]
On Fri, 27 Nov 2020, Mark Whitehorn wrote:

> +    logger.Write_MessageF("Frame: %s, Type: %s", motors->get_frame_string(), motors->get_type_string());
> 
> Wouldn't that conflict with a lot of user-facing stuff: https://ardupilot.org/copter/docs/frame-type-configuration.html?highlight=frame%20type

Yeah, you're right.  *sigh*.

**Comment by kd0aij** [comment]
@peterbarker 
> We're now calling methods on the motors object - we'd best see what happens if you don't have them defined properly - is it possible to get a method-on-undef problem?

That sounds like a 4GL thing... does making get_frame_string/type pure virtual address the issue?

**Comment by peterbarker** [comment]
On Sun, 6 Dec 2020, Mark Whitehorn wrote:

>       We're now calling methods on the motors object - we'd best see what happens if you don't have them defined properly - is it possible to get a
>       method-on-undef problem?
> 
> That sounds like a 4GL thing... does making get_frame_string/type pure virtual address the issue?

You're still potentially calling a method on nullptr - so pure-virtual 
isn't going to help you there.

(sorry, by "undef" I meant "nullptr" - perl definitely rots the brain ;-)  )

**Comment by kd0aij** [comment]
Added checks for nullptr... could happen in copter if gcs_mavlink starts before the motors object is allocated.
(didn't try to figure out whether that is possible)

**Comment by kd0aij** [comment]
This is failing CI on a Rover test, but I don't see how this PR could have any effect on Rover, since its motors class does not inherit from AP_Motors.

File: ArduCopter/GCS_Copter.cpp [hunk 10]
```diff
@@ -9,7 +9,11 @@ uint8_t GCS_Copter::sysid_this_mav() const
 
 const char* GCS_Copter::frame_string() const
 {
-    return copter.get_frame_string();
+    if (copter.motors == nullptr) {
+        return "motors not allocated";
+    } else {
```
Comment (peterbarker) [inline ArduCopter/GCS_Copter.cpp:14]: Don't need this else.

File: ArduCopter/GCS_Mavlink.cpp [hunk 11]
```diff
@@ -529,7 +529,13 @@ bool GCS_MAVLINK_Copter::params_ready() const
 void GCS_MAVLINK_Copter::send_banner()
 {
     GCS_MAVLINK::send_banner();
-    send_text(MAV_SEVERITY_INFO, "Frame: %s", copter.get_frame_string());
+    if (copter.motors == nullptr) {
+        send_text(MAV_SEVERITY_INFO, "motors not allocated");
+    } else {
```
Comment (peterbarker) [inline ArduCopter/GCS_Mavlink.cpp:534]: Early-return, please :-)

File: ArduCopter/Log.cpp (hunk 3 above)
Comment (peterbarker) [inline ArduCopter/Log.cpp:620]: Slightly different text string between this and the `send_text`.

File: ArduPlane/quadplane.cpp [hunk 12]
```diff
@@ -637,7 +636,9 @@ bool QuadPlane::setup(void)
       that the objects don't affect the vehicle unless enabled and
       also saves memory when not in use
      */
-    motor_class = (enum AP_Motors::motor_frame_class)frame_class.get();
+    motor_class = (AP_Motors::motor_frame_class)frame_class.get();
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:639]: We have a new toy we should use here.

Where we define `frame_class, which should change it to read:

```
AP_Enum<AP_Motors::motor_frame_class> frame_class;
```
After than you don't need the .get() *or* the cast :-)

similarly for _type

File: ArduPlane/quadplane.cpp [hunk 13]
```diff
@@ -658,12 +659,15 @@ bool QuadPlane::setup(void)
         SRV_Channels::set_default_function(CH_8, SRV_Channel::k_motor4);
         SRV_Channels::set_default_function(CH_11, SRV_Channel::k_motor7);
         AP_Param::set_frame_type_flags(AP_PARAM_FRAME_TRICOPTER);
+        coptar_tailsitter_supported = false;
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:662]: Does the deserve a different PR?  It's not related to moving strings around?

Also... `coptar`?!

File: ArduPlane/quadplane.cpp [hunk 14]
```diff
@@ -730,10 +738,10 @@ bool QuadPlane::setup(void)
     }
     AP_Param::load_object_from_eeprom(loiter_nav, loiter_nav->var_info);
 
-    motors->init((AP_Motors::motor_frame_class)frame_class.get(), (AP_Motors::motor_frame_type)frame_type.get());
+    motors->init(motor_class, motor_type);
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:741]: This is nice.

File: libraries/AP_Motors/AP_MotorsCoax.h [hunk 15]
```diff
@@ -49,6 +49,8 @@ class AP_MotorsCoax : public AP_MotorsMulticopter {
     //  this can be used to ensure other pwm outputs (i.e. for servos) do not conflict
     uint16_t            get_motor_mask() override;
 
+    const char* get_frame_string() override { return "COAX"; }
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_MotorsCoax.h:52]:
```suggestion
    const char* get_frame_string() const override { return "COAX"; }
```

... that will need to be done on all of them

File: libraries/AP_Motors/AP_MotorsMatrix.h [hunk 16]
```diff
@@ -55,6 +55,9 @@ class AP_MotorsMatrix : public AP_MotorsMulticopter {
     // using copter motors for forward flight
     float               get_roll_factor(uint8_t i) override { return _roll_factor[i]; }
 
+    const char*         get_frame_string() override { return _frame_class_string; }
+    const char*         get_type_string() override { return _frame_type_string; }
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_MotorsMatrix.h:59]:
```suggestion
    const char*         get_frame_string() const override { return _frame_class_string; }
    const char*         get_type_string() const override { return _frame_type_string; }
```

File: libraries/AP_Motors/AP_MotorsMatrix.h [hunk 17]
```diff
@@ -88,10 +91,14 @@ class AP_MotorsMatrix : public AP_MotorsMulticopter {
...
 
     // motor failure handling
     float               _thrust_rpyt_out_filt[AP_MOTORS_MAX_NUM_MOTORS];    // filtered thrust outputs with 1 second time constant
     uint8_t             _motor_lost_index;  // index number of the lost motor
+
+    motor_frame_class   _active_frame_class; // active frame class (i.e. quad, hexa, octa, etc)
+    motor_frame_type    _active_frame_type;  // active frame type (i.e. plus, x, v, etc)
+
+    const char*         _frame_class_string = ""; // string representation of frame class
+    const char*         _frame_type_string = "";  //  string representation of frame type
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_MotorsMatrix.h:103]: Maybe set these to UNKNOWN straight-up - or something distinctive?

File: libraries/AP_Motors/AP_Motors_Class.h [hunk 18]
```diff
@@ -64,6 +68,9 @@ class AP_Motors {
         MOTOR_FRAME_TYPE_BF_X_REV = 18, // X frame, betaflight ordering, reversed motors
     };
 
+    // return string corresponding to frame_type
+    virtual const char* get_type_string() { return ""; }
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_Motors_Class.h:72]: Perhaps return something distinctive here?

Also, if this is a can't-happen, please put in an internal_error::flow_of_control

**Review by peterbarker — CHANGES_REQUESTED** [review]
Looking good!

File: ArduCopter/GCS_Copter.cpp (hunk 10 above)
Comment (kd0aij) [reply ArduCopter/GCS_Copter.cpp:14]: fixed

File: ArduCopter/GCS_Mavlink.cpp (hunk 11 above)
Comment (kd0aij) [reply ArduCopter/GCS_Mavlink.cpp:534]: fixed

File: ArduPlane/quadplane.cpp (hunk 12 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:639]: needed a cast to use them in switch statements
and casting to unsigned gets an "ambiguous conversion" compiler error in the call to ::printf, but not in the call to config_error???
--separate commit for this one

File: ArduPlane/quadplane.cpp (hunk 13 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:662]: probably... I'll have to think about how to split those changes into a separate commit though
Don't like the term "copter tailsitter"? I'm not sure whether to blame me or @IamPete1 :)

File: ArduPlane/quadplane.cpp (hunk 14 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:741]: AP_Param is weird

File: libraries/AP_Motors/AP_Motors_Class.h (hunk 18 above)
Comment (kd0aij) [reply libraries/AP_Motors/AP_Motors_Class.h:72]: some frame classes don't have types, so an empty string is appropriate, but perhaps "unused" would be better?
If frame_type is significant, then this should be overridden with something more appropriate.
I'm not sure an internal error would be a good idea, since this is a public method and we have no control over who's calling it.

File: ArduCopter/Log.cpp (hunk 3 above)
Comment (kd0aij) [reply ArduCopter/Log.cpp]: fixed

File: libraries/AP_Motors/AP_MotorsMatrix.h (hunk 17 above)
Comment (kd0aij) [reply libraries/AP_Motors/AP_MotorsMatrix.h:103]: "uninitialized" would be accurate, but I think this wouldn't tell you anything useful since this just means the motors object failed to initialize for some reason...
Actually, for MotorsMatrix, these strings will be set to "UNSUPPORTED" appropriately when setup_motors fails and _intialized_ok returns false.

File: ArduPlane/quadplane.cpp (hunk 13 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: removed those changes and squashed

**Comment by peterbarker** [comment]
@kd0aij time to move this away from being a draft?

File: ArduCopter/GCS_Copter.cpp (hunk 1 above)
Comment (peterbarker) [reply ArduCopter/GCS_Copter.cpp:15]: No longer pointless.

File: ArduPlane/quadplane.cpp [hunk 19]
```diff
@@ -637,8 +634,7 @@ bool QuadPlane::setup(void)
       that the objects don't affect the vehicle unless enabled and
       also saves memory when not in use
      */
-    motor_class = (enum AP_Motors::motor_frame_class)frame_class.get();
-    switch (motor_class) {
+    switch ((AP_Motors::motor_frame_class)frame_class) {
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:637]:
```suggestion
    switch (frame_class) {
```

no cast required, kind of the point of `AP_Enum<>` :-)

File: ArduPlane/quadplane.cpp [hunk 20]
```diff
@@ -662,12 +658,13 @@ bool QuadPlane::setup(void)
...
     default:
-        AP_BoardConfig::config_error("Unknown Q_FRAME_CLASS %u", (unsigned)frame_class.get());
+        AP_BoardConfig::config_error("Unsupported Q_FRAME_CLASS %u", frame_class);
     }
+    ::printf("QuadPlane initialise, class: %u, type: %u\n", frame_class.get(), frame_type.get());
 
     if (tailsitter.motor_mask == 0) {
         // this is a normal quadplane
-        switch (motor_class) {
+        switch ((AP_Motors::motor_frame_class)frame_class) {
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:667]:
```suggestion
        switch (frame_class) {
```

Prettier without the cast :-)

File: ArduPlane/quadplane.cpp [hunk 21]
```diff
@@ -689,7 +686,7 @@ bool QuadPlane::setup(void)
         // this is a copter tailsitter with motor layout specified by frame_class and frame_type
         // tilting motors are not supported (tiltrotor control variables are ignored)
         if (tilt.tilt_mask != 0) {
-            gcs().send_text(MAV_SEVERITY_INFO, "Warning: Motor tilt not supported");
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:692]: To be clear, these send_text messages would never come out?

File: ArduPlane/quadplane.cpp [hunk 22]
```diff
@@ -730,10 +727,10 @@ bool QuadPlane::setup(void)
     }
     AP_Param::load_object_from_eeprom(loiter_nav, loiter_nav->var_info);
 
-    motors->init((AP_Motors::motor_frame_class)frame_class.get(), (AP_Motors::motor_frame_type)frame_type.get());
+    motors->init(frame_class, frame_type);
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:730]: Nice!

File: ArduPlane/quadplane.cpp [hunk 23]
```diff
@@ -730,10 +727,10 @@ bool QuadPlane::setup(void)
...
 
     if (!motors->initialised_ok()) {
-        AP_BoardConfig::config_error("unknown Q_FRAME_TYPE %u", (unsigned)frame_type.get());
+        AP_BoardConfig::config_error("init failed: class: %u type: %u", frame_class, frame_type);
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:733]: nominating the parameter name here (`Q_FRAME_TYPE`) is probably a good idea.
```suggestion
        AP_BoardConfig::config_error("init failed: Q_FRAME_CLASS=%u Q_FRAME_TYPE=%u", frame_class, frame_type);
```

File: ArduPlane/quadplane.cpp [hunk 24]
```diff
@@ -772,7 +769,7 @@ bool QuadPlane::setup(void)
     // param count will have changed
     AP_Param::invalidate_count();
 
-    gcs().send_text(MAV_SEVERITY_INFO, "QuadPlane initialised");
+    ::printf("QuadPlane initialised, class: %s, type: %s\n", motors->get_frame_string(), motors->get_type_string());
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:772]: This is a third variant of the way we're presenting the class/type.

I'm going to suggest a fourth:
`QUAD/PLUS`

just because.

File: libraries/AP_Motors/AP_Motors6DOF.cpp [hunk 25]
```diff
@@ -283,9 +289,9 @@ float AP_Motors6DOF::get_current_limit_max_throttle()
 // ToDo calculate headroom for rpy to be added for stabilization during full throttle/forward/lateral commands
 void AP_Motors6DOF::output_armed_stabilizing()
 {
-    if ((sub_frame_t)_last_frame_class == SUB_FRAME_VECTORED) {
+    if ((sub_frame_t)_active_frame_class == SUB_FRAME_VECTORED) {
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_Motors6DOF.cpp:292]: Could use `AP_Enum<sub_frame_t> _active_frame_class;`, but I'm pretty sure you've done enough cleanup on this PR :-)

**Review by peterbarker — CHANGES_REQUESTED** [review]
Still need the `const` on the frame-string-getters.

I'm still keen on changing `get_type_string` to get_layout_string` - but that's not your fight :-)

File: ArduPlane/quadplane.cpp (hunk 20 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:667]: but doesn't compile?

File: ArduPlane/quadplane.cpp (hunk 24 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:772]: I like that... what does @rmackay9 think?

File: libraries/AP_Motors/AP_MotorsCoax.h (hunk 15 above)
Comment (kd0aij) [reply libraries/AP_Motors/AP_MotorsCoax.h]: done

File: ArduPlane/quadplane.cpp (hunk 21 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:692]: right, I think QuadPlane::setup runs long before the connection to the gcs is active

**Comment by kd0aij** [comment]
put cast back: (AP_Motors::motor_frame_class)frame_class to avoid compilation error
and rebased on master to fix conflicts...I made several mistakes in the first pass at fixing conflicts; hope it's OK now

File: ArduCopter/system.cpp [hunk 26]
```diff
@@ -480,10 +480,47 @@ const char* Copter::get_frame_string() const
...
         case AP_Motors::MOTOR_FRAME_UNDEFINED:
-        default:
             return "UNKNOWN";
     }
+    // unknown frame
+    return "UNKNOWN";
+}
+
+// return MAV_TYPE corresponding to frame class
+MAV_TYPE Copter::get_frame_mav_type()
```
Comment (IamPete1) [inline ArduCopter/system.cpp:493]: I think this needs a rebase, I just moved it to motors in https://github.com/ArduPilot/ardupilot/commit/d0e1e85b9e43e181b597d94f8cf9de5515b079e5#diff-019889aa8f604ba16f535c5d61b768827e08bf3d1934b5e930f74b7645ae1405

Comment (kd0aij) [reply ArduCopter/system.cpp:493]: that's confusing... I though I rebased this morning on latest master

**Comment by kd0aij** [comment]
squashed and retested in SITL

File: libraries/AP_Motors/AP_MotorsMatrix.h [hunk 27]
```diff
@@ -70,6 +70,9 @@ class AP_MotorsMatrix : public AP_MotorsMulticopter {
     // using copter motors for forward flight
     float               get_roll_factor(uint8_t i) override { return _roll_factor[i]; }
 
+    const char*         get_frame_string() const override { return _frame_class_string; }
```
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsMatrix.h:73]: we should put a comment above this method if possible.

**Review by rmackay9 — APPROVED** [review]
Looks good, thanks!  Could we test Sub before merging that would be great.

**Review by peterbarker — APPROVED** [review]
Tested in Copter; works nicely.

Tested in Sub - doesn't have much effect in master, but https://github.com/ArduPilot/ardupilot/pull/16564 shows the frame class as part of the banner and it works as it should.

**Comment by peterbarker** [comment]
Merging this after agreement at DevCall - thanks @kd0aij !

Also, apologies for suggesting you try that Enum-Class thing - not quite working correctly with switch yet.
